content/sketch/part-1-the-engine.md

```diff
@@ -0,0 +1,243 @@
+```
+
+The generated file is located in `_build/default/one/engine.bc.js`, let's try that in Node REPL:
+
+```bash
+❯ node
+> const { execute } = require("./_build/default/one/engine.bc.js")
+undefined
+
+> execute("1 + 1;;")
+'- : int = 2\n'
+```
+
+Great! So we got ourselves a usable engine that you can embed in your website or calling it from Node.js. But this naive version of the engine has some limitations:
+
+- Errors and logs (stdout, stderr in OCaml) is printed directly to Javascript console, we want to catch them and return them as result of `execute` function for displaying in the UI.
+
+```bash
+> execute(`print_endline "Hello world from OCaml";;`);
+Hello world from OCaml
+'- : unit = ()\n'
+
+> execute(`syntax error`);
+File "", line 1, characters 12-12:
+Error: Syntax error
+''
+```
+
+- It executes all code at once. We want it to return the result for each individual statement (they are called phrases in OCaml's term).
+
+```bash
+> execute(`let add = (+);; add 1 2;; add 3 4;;`);
+'val add : int -> int -> int = <fun>\n- : int = 3\n- : int = 7\n'
+```
+
+- It doesn't support ReasonML syntax yet.
+
+# Turn `stdout` and `stderr` into values
+
+`stdout` and `stderr` are `out_channel`. They are flushed to Javscript console by default but we can flush them into a buffer instead.
+
+```reason
+let stderr_buffer = Buffer.create(100);
+let stdout_buffer = Buffer.create(100);
+
+Sys_js.set_channel_flusher(stdout, Buffer.add_string(stdout_buffer));
+Sys_js.set_channel_flusher(stderr, Buffer.add_string(stderr_buffer));
+```
+
+Add these lines to the end of `execute` function: 
+
+```reason
+let result = Buffer.contents(buffer);
+let stderr_result = Buffer.contents(stderr_buffer);
+let stdout_result = Buffer.contents(stdout_buffer);
+
+Buffer.clear(stderr_buffer);
+Buffer.clear(stdout_buffer);
+
+[%js {
+  val result = Js.string(result);
+  val stderr = Js.string(stderr_result);
+  val stdout = Js.string(stdout_result)
+}]
+```
+
+Here we bind the content of execution result, stderr and stdout into variables and clear the buffer after that. We also return an Javascript object with this shape:
+
+```js
+{
+  result: string,
+  stderr: string,
+  stdout: string
+}
+```
+
+We can build this example with:
+
+```
+esy dune build two/engine.bc.js --profile release
+```
+
+Test it:
+
+```
+❯ node
+> const { execute } = require("./_build/default/two/engine.bc.js")
+undefined
+
+> execute("1 + 1;;")
+{ result: '- : int = 2\n', stderr: '', stdout: '' }
+
+> execute(`print_endline "Hello world from OCaml";;`);
+{ result: '- : unit = ()\n',
+  stderr: '',
+  stdout: 'Hello world from OCaml\n' }
+
+> execute(`syntax error`);
+{ result: '',
+  stderr: 'File "", line 1, characters 12-12:\nError: Syntax error\n',
+  stdout: '' }
+```
+
+That looks more promising! We get `stderr` and `stdout` as return values. In Sketch UI, This information will be display inline like this
+
+![Sketch inline errors and values](./images/display-inline.png)
+
```

Comment:
What about a final recap or something like "in the next post, you will learn about ..."?